src/OpenTelemetry/Metrics/AggregatorStore.cs

```diff
@@ -178,27 +179,46 @@ private void InitializeZeroTagPointIfNotInitialized()
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         private int LookupAggregatorStore(string[] tagKeys, object[] tagValues, int length)
         {
-            int aggregatorIndex;
-            string[] seqKey = null;
+            var givenTags = new Tags(tagKeys, tagValues);
 
-            // GetOrAdd by TagKeys at 1st Level of 2-level dictionary structure.
-            // Get back a Dictionary of [ Values x Metrics[] ].
-            if (!this.keyValue2MetricAggs.TryGetValue(tagKeys, out var value2metrics))
+            // We only need to sort if there is more than one Tag Key.
+            if (!this.sortedTagsDictionary.TryGetValue(givenTags, out var sortedTags))
             {
                 // Note: We are using storage from ThreadStatic, so need to make a deep copy for Dictionary storage.
-                seqKey = new string[length];
-                tagKeys.CopyTo(seqKey, 0);
+                var givenKeys = new string[length];
+                tagKeys.CopyTo(givenKeys, 0);
+
+                var givenValues = new object[length];
+                tagValues.CopyTo(givenValues, 0);
+
+                givenTags = new Tags(givenKeys, givenValues);
+
+                string[] sortedTagKeys;
+                object[] sortedTagValues;
+
+                if (length > 1)
+                {
+                    // Create a new array for the sorted Tag keys.
+                    sortedTagKeys = new string[length];
+                    tagKeys.CopyTo(sortedTagKeys, 0);
+
+                    // Create a new array for the sorted Tag values.
+                    sortedTagValues = new object[length];
+                    tagValues.CopyTo(sortedTagValues, 0);
 
-                value2metrics = new ConcurrentDictionary<object[], int>(ObjectArrayComparer);
-                if (!this.keyValue2MetricAggs.TryAdd(seqKey, value2metrics))
+                    Array.Sort(sortedTagKeys, sortedTagValues);
+                }
+                else
                 {
-                    this.keyValue2MetricAggs.TryGetValue(seqKey, out value2metrics);
+                    sortedTagKeys = givenKeys;
+                    sortedTagValues = givenValues;
                 }
+
+                sortedTags = new Tags(sortedTagKeys, sortedTagValues);
+                this.sortedTagsDictionary.TryAdd(givenTags, sortedTags);
```

Comment:
Similar to comment here https://github.com/open-telemetry/opentelemetry-dotnet/pull/2777#discussion_r783563923

Can we defer this add until after `this.tagsToMetricPointIndexDictionary.TryAdd(sortedTags, aggregatorIndex);` below? I figure at we at least know we've gotten a metric point by then.

My concern is that one poorly instrumented library that suffers from a dimensional explosion could cause this dictionary to grow unbounded.

Reply:
We cannot defer it as we use the value of the `sortedTagsDictionary` as a key to lookup the `tagsToMetricPointIndexDictionary`.

We could add this check right after `if (!this.sortedTagsDictionary.TryGetValue(givenTags, out var sortedTags))`
```
if (!this.sortedTagsDictionary.TryGetValue(givenTags, out var sortedTags))
{
    aggregatorIndex = this.metricPointIndex;
    if (aggregatorIndex >= this.maxMetricPoints)
    {
        // sorry! out of data points.
        // TODO: Once we support cleanup of
        // unused points (typically with delta)
        // we can re-claim them here.
        return -1;
    }
    ...
    ...
    this.sortedTagsDictionary.TryAdd(givenTags, sortedTags);
}
```

This way we could avoid adding new entries to `sortedTagsDictionary` once we run out of MetricPoints. Does that seem good?

Reply:
This would still not completely prevent the dictionary from growing too large. If there is a bad instrumentation that emitsall possible combinations of a set of tags, they would all contribute to just a single `MetricPoint` but we would still have an entry for each of those combinations.

For e.g. If we have a Counter with 8 tags. #possible combinations = 8! = 40320. If the code is instrumented badly such that all these combinations are emitted then we would have 40320 entries in the `sortedTagsDictionary` which would all be contributing to just one `MetricPoint`.

Reply:
> This would still not completely prevent the dictionary from growing too large. If there is a bad instrumentation that emitsall possible combinations of a set of tags, they would all contribute to just a single MetricPoint but we would still have an entry for each of those combinations

^ We need to address this. We enforce MaxMetrics and MaxMetricPointPerMetric caps to bound the resource usage. this negates them and allows unbound growth. We could store only the unsorted->sorted for the very 1st time, and not after. if user keeps providing different order, we'll just sort it - its slow, but won't lead to unbounded memory

Reply:
> We could store only the unsorted->sorted for the very 1st time, and not after.

I'm not sure this solves the issue. @utpilla reminded me it's not just about combinations of sort order. It's also about the combinations of values.

Assume a metric with the following dimensions and range of value for each dimension.
```
A = [red, black, blue], B = [sour, tangy, sweet], C = [gross, delicious]
```

We'll need to store 18 different combinations for this in `sortedTagsDictionary` - and this example is pretty reasonable cardinality-wise. Though, in poorly designed instrumentation, one of these dimensions may be unbounded. Imagine an `http.url` attribute where `url` was infinite. I've seen something like this before in pen tests where an application is hit with a barrage of bogus urls. If those urls are used as a value in one of the metric dimensions then... bad news bears 🐻 - the OTel SDK just took your app down.

Maybe we just need to pick another arbitrary cap on the size of the `sortedTagsDictionary` and then adjust it when we learn more.

Reply:
@alanwest see this -> https://github.com/open-telemetry/opentelemetry-dotnet/pull/2805#discussion_r793795826 It might solve the unbounded memory usage issue..

Reply:
Ah ha, yes, didn't see your suggestion earlier.